### PR TITLE
Issue 31+32

### PR DESCRIPTION
**Pull request by ememisya**
**What's New?**
- Refactored asn package to decoder (as it does more than ASN now)
- Added decoder for SSH2 RSA & DSA public keys (RFC-4253)
- Added PemReader which validates and parses PEMs based on their respective RFCs
- New utils to help with the features above

### CODE REVIEW

**Comment by ememisya** [comment]
Given there are API changes (major release), would you like for me to refactor so that I add the decoder package and leave asn alone for now, and perhaps we deprecate it and merge in the next major version release?  Or would you rather we wait to merge this into the next major release?  Or perhaps this merits a new major release?

File: src/main/java/org/cryptacular/decoder/PublicKeyDecoder.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    try {
+      String encodedString = new String(encoded, 0, 10, ByteUtil.ASCII_CHARSET).trim();
+      if (PemUtil.isRFC4716Pem(encoded)) {
+        returnValue =  new SSH2PublicKeyDecoder().decode(PemUtil.decodeToPem(encoded).getContent());
+      } else if (encodedString.startsWith("ssh-")) {
+        encodedString = new String(encoded, 0, encoded.length, ByteUtil.ASCII_CHARSET).trim();
+        returnValue =  new SSH2PublicKeyDecoder().decode(
+                CodecUtil.b64(encodedString.substring(
+                        encodedString.indexOf(" ") + 1, encodedString.lastIndexOf(" ") + 1)
+                )
```
Comment (serac) [inline src/main/java/org/cryptacular/decoder/PublicKeyDecoder.java]: We don't use this style of closing parentheses in our Java projects.

File: src/main/java/org/cryptacular/decoder/SSH2PublicKeyDecoder.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  /**
+   * Decodes a Java String from a string as per RFC 4251 Section 5 using the current
+   * position of the bytes buffer
+   * @param bytes buffered bytes array to read from
+   * @return {@link String} representing string
+   */
+  private String decodeString(final byte[] bytes)
+  {
+    int length = decodeUInt32(bytes);
```
Comment (serac) [inline src/main/java/org/cryptacular/decoder/SSH2PublicKeyDecoder.java]: Please use ByteUtil.toInt() and add an overloaded method signature to take an offset:

    int length = ByteUtil.toInt(bytes, position);
    position += 4;

File: src/main/java/org/cryptacular/io/pem/Pem.java [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+import java.util.stream.Stream;
+import org.bouncycastle.util.io.pem.PemHeader;
+import org.bouncycastle.util.io.pem.PemObject;
+
+/**
+ * Container for PEM encoded data.
+ *
+ * @author Middleware Services
+ */
+public final class Pem extends PemObject
```
Comment (serac) [inline src/main/java/org/cryptacular/io/pem/Pem.java]: Needs more descriptive name. Suggest `EncapsulatedPemObject`.

File: src/main/java/org/cryptacular/io/pem/PemReader.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+      }
+      base64DataBuilder.append(line.trim());
+    }
+    final String b64buffer = base64DataBuilder.toString();
+    if (line == null) {
+      throw new IOException(endMarker + " not found");
+    }
+    enforceLineLengthRestrictions(rfcFormat, lineLength);
+    if (rfcFormat.equals(Pem.Format.RFC7468)) {
+      return new Pem(type, Base64.decode(b64buffer), explanatoryText);
```
Comment (serac) [inline src/main/java/org/cryptacular/io/pem/PemReader.java]: Please use Cryptacular encoding routines:

     return new Pem(type, CodecUtil.b64(b64Buffer), explanatoryText);

**Comment by serac** [comment]
Looks good generally, but I'm concerned about moving existing components from the `asn` package to `decoder`. I don't think these changes merit a major version bump, so let's go with duplicating and deprecating for now.

**Comment by ememisya** [comment]
Pushed a4dfda3 with the changes.

**Comment by ememisya** [comment]
Pushed 4b09660 and dae85ea with the changes discussed.

**Comment by ememisya** [comment]
Any update on this?  It's been a while :smile:

**Comment by serac** [comment]
I have a slight concern about effectively changing package names, but I think we've handled it fairly well via deprecation. I think that's relatively low cost for the functionality we get. @dfish3r do you have any concerns about accepting this work?

**Comment by ememisya** [comment]
Hi @dfish3r, just checking in on this branch.  Any plans to move forward with the work?

**Comment by ememisya** [comment]
Hey guys, any news on this?  Looks like it hasn't strayed far behind despite the time elapsed.

**Comment by ememisya** [comment]
So why wasn't this implemented?

**Pull request closed without merging**